libraries/reconstruction_cuda/cuda_gpu_iirconvolve.cu

```diff
@@ -1,107 +1,314 @@
+    const float z = sqrtf(3.f) - 2.f;
+    const float k = z / (z - 1.f);
+    const float gain = 6.0f;
 
-__global__
-void iirConvolve2D_Cardinal_Bspline_3_MirrorOffBound(float* input,
-        float* output, size_t xDim, size_t yDim) {
-    // assign line to thread
-    int idy = blockIdx.y * blockDim.y + threadIdx.y;
-    if (idy >= yDim) return; // only threads with data should continue
-    float* line = output + (idy * xDim);
-
-    // adjust gain
-    float z = sqrtf(3.f) - 2.f;
-    float z1 = 1.0 - z;
-    float gain = -(z1 * z1) / z;
-
-    // copy original data
-    for (int i = 0; i < xDim; i++) {
-        line[i] = input[i + (idy * xDim)] * gain;
-    }
-    // prepare some values
-    float sum = (line[0] + powf(z, xDim) * line[xDim - 1]) * (1.f + z) / z;
-    z1 = z;
-    float z2 = powf(z, 2 * xDim - 2);
-    float iz = 1.f / z;
-    for (int j = 1; j < (xDim - 1); ++j) {
-        sum += (z2 + z1) * line[j];
-        z1 *= z;
-        z2 *= iz;
-    }
-    line[0] = sum * z / (1.f - powf(z, 2 * xDim));
-    for (int j = 1; j < xDim; ++j) {
-        line[j] += z * line[j - 1];
-    }
-    line[xDim - 1] *= z / (z - 1.f);
-    for (int j = xDim - 2; 0 <= j; --j) {
-        line[j] = z * (line[j + 1] - line[j]);
+    const int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float *myCol = in + col;
+
+    if (col < x) {
+
+        myCol[0] *= z * gain;
+        #pragma unroll UNROLL
+        for (int j = 1; j < y; ++j) {
+            myCol[j*x] = myCol[j*x] * gain + z * myCol[(j - 1)*x];
+        }
+
+        myCol[(y - 1)*x] *= k;
+        #pragma unroll UNROLL
+        for (int j = y - 2; 0 <= j; --j) {
+            myCol[j*x] = z * (myCol[(j + 1)*x] - myCol[j*x]);
+        }
     }
 }
+
+/*
+ * Faster when compiled with fixed `x` and `y`
+ * speed on GeForce GTX 1070 on size 8192x8192:
+    a) fixed size: 6100 MPix/s
+    b) variable size: 5100 MPix/s
+*/
+void solveGPU(float *in, int x, int y) {
+
+    const int x_padded = x / COL_BLOCK_SIZE * COL_BLOCK_SIZE + COL_BLOCK_SIZE * (x % COL_BLOCK_SIZE != 0);
+    const int y_padded = y / BLOCK_SIZE * BLOCK_SIZE + BLOCK_SIZE * (y % BLOCK_SIZE != 0);
+
+    sum_lines<<<y, WARP_SIZE>>>(in, x);
+    gpuErrchk(cudaPeekAtLastError());
+    gpuErrchk(cudaDeviceSynchronize());
+
+    rows_iterate<<<y_padded / BLOCK_SIZE, BLOCK_SIZE>>>(in, x);
+    gpuErrchk(cudaPeekAtLastError());
+    gpuErrchk(cudaDeviceSynchronize());
+
+    sum_cols<<<x, WARP_SIZE>>>(in, x);
+    gpuErrchk(cudaPeekAtLastError());
+    gpuErrchk(cudaDeviceSynchronize());
+
+    cols_iterate<<<x_padded / COL_BLOCK_SIZE, COL_BLOCK_SIZE>>>(in, x, y);
+    gpuErrchk(cudaPeekAtLastError());
+    gpuErrchk(cudaDeviceSynchronize());
+}
+
+} // namespace iirConvolve2D_Cardinal_BSpline_3_MirrorOffBoundKernels
+
+void iirConvolve2D_Cardinal_Bspline_3_MirrorOffBound(float* input,
+        float* output, int xDim, int yDim) {
+    iirConvolve2D_Cardinal_BSpline_3_MirrorOffBoundKernels::solveGPU( input, xDim, yDim );
```

Comment:
where is the output?

Reply:
Kernels does the transformation inplace. I changed arguments of the function to make explicit that it doesn't use any other memory.